File: Algorithms/24. Swap Nodes in Pairs/swift/solution.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func swapPairs(_ head: ListNode?) -> ListNode? {
        var (cur, next) = (head, head?.next)
        var answer = next ?? cur

        while cur != nil, next != nil {
            var _cur = next?.next
            var _next = _cur?.next

            next?.next = cur
            cur?.next = _next ?? _cur

            (cur, next) = (_cur, _next)
        }

        return answer
    }
}

File: Algorithms/24. Swap Nodes in Pairs/kotlin/solution.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun swapPairs(head: ListNode?): ListNode? {
        if (head == null || head.next == null) {
            return head
        }

        val dummy = ListNode(0)

        var current: ListNode? = head
        var prevNode: ListNode? = dummy

        while (current != null && current?.next != null) {
            val secondNext = current?.next?.next
            val second = current?.next

            prevNode?.next = second
            second?.next = current
            current?.next = secondNext

            prevNode = prevNode?.next?.next
            current = secondNext
        }


        return dummy.next
    }
}
